cli: rework -addrinfo cli to use addresses which aren’t filtered for quality/recency

stratospher opened the pull request:
Rework of `-addrinfo` CLI is done using `getaddrmaninfo` RPC proposed in https://github.com/bitcoin/bitcoin/pull/27511. This would be useful for users who want to know the total number of addresses the node knows about and can make connections to. 

Currently, `-addrinfo` returns total number of addresses the node knows about after filtering them for quality + recency using [`isTerrible`](https://github.com/bitcoin/bitcoin/blob/4b51290f719559b8506425177f7758abe66fe2c3/src/addrman.cpp#L808). However `isTerrible`addresses [don't matter](https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1147725684) when selecting outbound peers to connect to. Total number of addresses the node knows about could be higher than what `-addrinfo` currently displays. See https://github.com/bitcoin/bitcoin/pull/24370.

stratospher committed:
cli: modify -addrinfo to use getaddrmaninfo RPC endpoint

currently -addrinfo returns addresses known to the node after
filtering for quality and recency. However, the node considers
all known addresses (even the filtered out addresses) when
selecting peers to connect to.

So update -addrinfo to also display the full set of known
addresses for more useful node information.

mzumsande reviewed (commented):
Concept ACK, will review soon.

amitiuttarwar commented:
obvious concept ack considering I proposed #26907

mzumsande commented on src/bitcoin-cli.cpp:
```diff
@@ -252,46 +252,30 @@ class BaseRequestHandler
         if (!reply["error"].isNull()) return reply;
-        const std::vector<UniValue>& nodes{reply["result"].getValues()};
-        if (!nodes.empty() && nodes.at(0)["network"].isNull()) {
+        const std::vector<UniValue>& addrman{reply["result"].getValues()};
```
I'd prefer a different name (maybe `addr_counts` or `addrman_counts`) to avoid confusion that the object isn't addrman, but the result of a query to addrman.

stratospher replied:
that's true. i've updated the PR.

mzumsande reviewed (commented):
Looks good to me, only minor nits.

ping @jonatack, could you have a look, especially at the changes to bitcoin-cli?

mzumsande commented on test/functional/rpc_net.py:
```diff
@@ -333,6 +334,20 @@ def test_addpeeraddress(self):
+
+        # current ipv4 count in node's Addrman: new 1, tried 1
+        self.log.debug("Test that addrmaninfo provides correct new/tried table address count")
+        node.addpeeraddress(address="3.1.1.1", tried=True, port=8333)
```
Maybe it would actually make sense to add the possibility of a deterministic addrman as, it e.g. we could possibly extend the existing `-addrmantest` test-only RPC. But no need to do it in this PR.

stratospher replied:
interesting, thanks! so that's why the ci was failing. i've updated the PR to reflect this.

mzumsande replied:
I don't think so - even with the guarantee that `GetGroup()` is different, there is still a small chance that two addresses happen to hash to the same bucket in `GetNewBucket()`.

stratospher replied:
discussed this more with @amitiuttarwar and understood that `nKey` would remain the [same](https://github.com/bitcoin/bitcoin/blob/8c4958bd4c06026dc108bc7f5f063d1f389d279b/src/addrman.cpp#L109) for a node's addrman and hence same bucket position.

i tried generating random ipv6 addresses and inserting into addrman:
```
while len(node.getnodeaddresses(count=0)) < 3:
            ipv6_address = ":".join(format(randint(0, 0xffff), '04x') for _ in range(8))
            node.addpeeraddress(address=ipv6_address, tried=False, port=8333)
```

https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1134455743 would still be a concern since there's a small chance that an ipv6 address could hash to the same bucket as an ipv4 address. this would cause the assertions to check the count of different network addresses in the test to fail.

so leaving out the network coverage tests for now.

0xB10C replied:
fyi: `addpeeraddress` returns `success` indicating if inserting was successful. I'm using this to retry inserting in case there was a collision.

```
added = False
while not added:
    result = node.addpeeraddress(address="..", tried=to_tried, port=port)
    added = result["success"]
```

stratospher commented on src/rpc/net.cpp:
```diff
@@ -974,6 +974,60 @@ static RPCHelpMan addpeeraddress()
+static RPCHelpMan addrmaninfo()
+{
+    return RPCHelpMan{"addrmaninfo",
+                      "\nReturns count of addresses in new/tried table of all networks(or a given network). This RPC is for testing only.\n",
```
done.

amitiuttarwar commented on src/rpc/net.cpp:
```diff
@@ -974,6 +974,60 @@ static RPCHelpMan addpeeraddress()
     };
 }
 
+static RPCHelpMan addrmaninfo()
```
realized it probably makes more sense to call it `getaddrmaninfo` to be consistent with other RPC messages

stratospher replied:
true, i've updated the PR to call it `getaddrmaninfo`.

brunoerg reviewed (commented):
Concept ACK

brunoerg commented on src/rpc/net.cpp:
```diff
@@ -966,6 +966,60 @@ static RPCHelpMan addpeeraddress()
+
+                          const std::optional<Network> network{request.params[0].isNull() ? std::nullopt : std::optional<Network>{ParseNetwork(request.params[0].get_str())}};
+                          if (network == NET_UNROUTABLE) {
+                              throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Network not recognized: %s", request.params[0].get_str()));
```
You could cover this error in the functional test

stratospher replied:
thanks! done.

brunoerg commented:
```sh
"network" : "str",    (string) The network (ipv4, ipv6, onion, i2p, cjdns)
```

Instead of being a string, wouldn't make sense it to be an array? E.g. I want to get new/tried table address count for ipv5 and ipv6 together.

stratospher commented:
>Instead of being a string, wouldn't make sense it to be an array? E.g. I want to get new/tried table address count for ipv5 and ipv6 together.

i'm confused about this since making it into `RPCArg::Type::ARR` would make the RPC harder to type and use? 
 `$ bitcoin-cli getaddrmaninfo "[\"ipv4\", \"ipv6\"]"`

and there are only few network types. would be interested in what others think.

brunoerg commented:
that was just a question, thinking about complexity I agree on leaving it as is.

brunoerg commented:
perhaps we could also check this is a hidden RPC like `addpeerinfo`?

```py
self.log.debug("Test that addpeerinfo is a hidden RPC")
# It is hidden from general help, but its detailed help may be called directly.
assert "addpeerinfo" not in node.help()
assert "addpeerinfo" in node.help("addpeerinfo")
```

stratospher committed:
doc: add release notes for #26988

stratospher commented:
thank you @amitiuttarwar and @brunoerg! I've updated the PR to include your suggestions:
- added release notes
- test coverage for different networks
- hidden RPC check

stratospher commented on test/functional/rpc_net.py:
```diff
@@ -328,6 +329,15 @@ def test_addpeeraddress(self):
             addrs = node.getnodeaddresses(count=0)  # getnodeaddresses re-runs the addrman checks
             assert_equal(len(addrs), 2)
 
+    def test_getaddrmaninfo(self):
```
done.

mzumsande commented on doc/release-notes/release-notes-26988.md:
```diff
@@ -0,0 +1,11 @@
+Tools and Utilities
+--------
+
+- CLI -addrinfo would return known addresses including `isTerrrible` addresses
+  now since it has been reworked to utilise `getaddrmaninfo` RPC. (#26988)
```
As far as I know, test-only RPCs, similar to debug-only bitcoind options, aren't usually mentioned in  release notes - the target group for release notes are regular users, while test-only options are meant for devs and may not be stable. Related test-only RPCs such as `addpeeraddress` weren't introduced in release notes either. So maybe only describe the changed behavior of cli -addrinfo and don't mention `getaddrmaninfo`?

stratospher replied:
> hm, seems a little low-level for release notes. maybe something like

sounds much better! thank you!

>  the target group for release notes are regular users, while test-only options are meant for devs and may not be stable.

interesting, thanks! i've updated it.

fanquake replied:
Seems kind of weird to change the CLI to use a "test-only" RPC in any case

stratospher replied:
>  Seems kind of weird to change the CLI to use a "test-only" RPC in any case

@fanquake, @amitiuttarwar `-generate` CLI currently [uses](https://github.com/bitcoin/bitcoin/blob/467fa8943801911c233cb96d45282b1de10bfa90/src/bitcoin-cli.cpp#L1178) a [hidden RPC](https://github.com/bitcoin/bitcoin/blob/467fa8943801911c233cb96d45282b1de10bfa90/src/rpc/mining.cpp#L1055) (`generatetoaddress`) too. 
i don't think a normal user would be interested in the new/tried table breakdown of addresses. so hopefully it's ok to keep it hidden.

stratospher replied:
> Maybe borrow from the current -addrinfo help, along the lines of CLI -addrinfo previously returned addresses known to the node after filtering for quality and recency. It now returns all of the addresses known to the node.

thanks @jonatack! this sounds much better.

> It might be interesting to have both with and without IsTerrible() filtering to maintain the current info returned and to be able to compare over time, either by returning a breakdown in the new RPC or by leaving -addrinfo unchanged to compare the two.

true! decided to keep the RPC/CLI in a separate PRs.

stratospher commented on test/functional/rpc_net.py:
```diff
@@ -328,6 +329,37 @@ def test_addpeeraddress(self):
+        node.addpeeraddress(address="pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion", tried=True, port=8333) # onion address
+        node.addpeeraddress(address="c4gfnttsuwqomiygupdqqqyy5y5emnk5c73hrfvatri67prd7vyq.b32.i2p", tried=False, port=8333)  # i2p address
+        node.addpeeraddress(address="1233:3432:2434:2343:3234:2345:6546:4534", tried=False, port=8333)                       # ipv6 address
+        node.addpeeraddress(address="fc00:1:2:3:4:5:6:7", tried=False, port=8333)                                            # cjdns address
```
makes sense, sporadic failures do happen in that commit (removed it now).
continued this discussion here - https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1137596851

brunoerg commented on test/functional/rpc_net.py:
```diff
@@ -290,7 +291,7 @@ def test_addpeeraddress(self):
         """
         self.log.info("Test addpeeraddress")
-        self.restart_node(1, ["-checkaddrman=1"])
+        self.restart_node(1, ["-checkaddrman=1", "-cjdnsreachable"])
```
What's the reason for `-cjdnsreachable` here?

stratospher replied:
oops, forgot to revert that. thanks!

brunoerg replied:
ohh, got it! thank you!

mzumsande commented on src/bitcoin-cli.cpp:
```diff
@@ -255,46 +255,30 @@ class BaseRequestHandler
         }
-        UniValue params{RPCConvertValues("getnodeaddresses", std::vector<std::string>{{"0"}})};
-        return JSONRPCRequestObj("getnodeaddresses", params, 1);
+        return JSONRPCRequestObj("getaddrmaninfo", NullUniValue, 1);
```
> how important are IsTerrible addresses, and would it be interesting to have the breakdown with/without that filtering.

`IsTerrible` is used for two things outside of RPCs:
1.) Terrible addresses are not included in answers to GETADDR requests from peers
2.) In case of a conflict (two addrs would go to the same bucket and position in the new tables) the old addr is only overwritten if it's terrible.

Notably, `IsTerrible` isn't used as a filter while making outbound connections, which is the primary function of addrman and the thing users would be most concerned about. That's why I think that the total number of addresses is the better number to report in RPCs.

stratospher replied:
noted. not convinced how desirable the additional code complexity is. would like to hear how other people think about code complexity too before deciding what to do next.

ajtowns commented:
My opinion only, but it seems like it would be simpler to have the RPC not take any arguments and always output an object:

```
$ bitcoin-cli getaddrinfo
{
    "ipv4": {
        "new": 5,
        "tried": 3
    },
    ...
}
```

If you want a specific network, you can use jq: `bitcoin-cli getaddrinfo | jq .ipv4` ? You can also do sums this way if desired: `| jq .ipv4.new + ipv6.new`. (Having an object rather than an array makes it easy to get at the values rather than an array and having to select by one of the fields)

stratospher commented on src/rpc/net.cpp:
```diff
@@ -966,6 +966,60 @@ static RPCHelpMan addpeeraddress()
+static RPCHelpMan getaddrmaninfo()
+{
+    return RPCHelpMan{"getaddrmaninfo",
+                      "\nReturns count of addresses in new/tried table of all networks (or a given network). This RPC is for testing only.\n",
```
you're right! I've updated `getaddrmaninfo` taking both these suggestions.

stratospher commented:
thank you for the useful feedback! Splitting this into 2 separate PRs since it’d be easier to think about RPC and CLI parts separately. I’ve opened https://github.com/bitcoin/bitcoin/pull/27511 for getaddrmaninfo RPC and updated this PR to reflect just the CLI changes.

- I liked returning objects as output and displaying total addresses per network ideas. Updated getaddrmaninfo in https://github.com/bitcoin/bitcoin/pull/27511 to use these.
- Haven't changed the previous CLI approach because of https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1173966799.

luke-jr commented:
>Rework of -addrinfo CLI is done using getaddrmaninfo RPC proposed in https://github.com/bitcoin/bitcoin/pull/27511.

If it's going to be used for this, it shouldn't be hidden/test-only...

ajtowns commented:
> it was kept as a hidden RPC because:
> 
>     1. a normal user wouldn't be interested in the new/tried table breakdown of addresses.
> 
>     2. `-generate CLI` currently [uses](https://github.com/bitcoin/bitcoin/blob/467fa8943801911c233cb96d45282b1de10bfa90/src/bitcoin-cli.cpp#L1178) a [hidden RPC](https://github.com/bitcoin/bitcoin/blob/467fa8943801911c233cb96d45282b1de10bfa90/src/rpc/mining.cpp#L1055) (generatetoaddress) too.

I think it makes sense to unhide this RPC; we have plenty of `get*info` functions that are only really useful experts, and that's fine. It makes sense to hide rpcs that are only useful for development (generate, setmocktime, echo) or that might cause the node to not work as expected (invalidateblock, sendmsgtopeer), but this is likely useful for some regular users and doesn't do anything potentially harmful, so there shouldn't be any reason to hide it IMO.

stratospher commented:
Rebased.

> I think it makes sense to unhide this RPC; we have plenty of get*info functions that are only really useful experts, and that's fine. It makes sense to hide rpcs that are only useful for development (generate, setmocktime, echo) or that might cause the node to not work as expected (invalidateblock, sendmsgtopeer), but this is likely useful for some regular users and doesn't do anything potentially harmful, so there shouldn't be any reason to hide it IMO.

That sounds reasonable. Done in #28565.

pablomartin4btc reviewed (commented):
tested ACK 0b4ccbd0358d95632e2e46bfdf7715b199971754

<details>
<summary>output before this PR</summary>

```
/src/bitcoin-cli -signet -addrinfo
{
  "addresses_known": {
    "ipv4": 1097,
    "ipv6": 352,
    "onion": 0,
    "i2p": 0,
    "cjdns": 0,
    "total": 1449
  }
}

```
</details>
<details>
<summary>output after this PR</summary>

```
./src/bitcoin-cli -signet -addrinfo
{
  "addresses_known": {
    "ipv4": 2157,
    "ipv6": 834,
    "onion": 0,
    "i2p": 0,
    "cjdns": 0,
    "all_networks": 2991,
    "total": 5982
  }
}

```
</details>

I share the same [concerns](https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1145326660) as @jonatack regarding maintaining backward compatibility.

Also, regarding both the totals and per network differences between before and after this changes I wonder if we should add some context in the output or in the help for the user to understand where those differences are coming from. (Mainly due to @jonatack's comment [here](https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1146970653) and @mzumsande's comment [here](https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1147725684) regarding `isTerrible()`).

Last thing for now, @stratospher I see you agreed with @brunoerg's [suggestion](https://github.com/bitcoin/bitcoin/pull/26988#issuecomment-1464036406) but I don't see the changes, perhaps it was removed and I missed some discussion about it.

sr-gi commented:
What is the state of this @stratospher? Are you still looking for feedback on https://github.com/bitcoin/bitcoin/pull/26988#issuecomment-1548215577?

stratospher commented on doc/release-notes/release-notes-26988.md:
```diff
@@ -0,0 +1,5 @@
+Tools and Utilities
+--------
+
+- CLI -addrinfo previously returned addresses known to the node after filtering for quality and recency.
```
liked it, done.

stratospher commented:
Thank you for the reviews @pablomartin4btc! I've rebased the PR and included your [suggestion](https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1360075599).

> Also, regarding both the totals and per network differences between before and after this changes I wonder if we should add some context in the output or in the help for the user to understand where those differences are coming from.

i prefer not adding niche details in the user output/help because the difference would only be felt temporarily when users upgrade and it is covered in the release notes.

> Last thing for now, @stratospher I see you agreed with @brunoerg's suggestion but I don't see the changes, perhaps it was removed and I missed some discussion about it.

yes! that suggestion isn't applicable anymore since the RPC is public. this was done in #27511.

stratospher commented:
Rebased. thanks @jonatack! reaching out.

pablomartin4btc reviewed (commented):
tACK c2f3225622cbbe516ac23059bff47ed3c28653cf

> my personal preference is for displaying only the relevant total addresses stats 

I also lean toward showing only the total known addresses, as I think it better reflects what the node has in its addrman and avoids suggesting that only the filtered set "counts."
I'd consider whether passing an argument might offer one or both views (filtered/unfiltered) depending on the use case, but I’ve read the previous discussions here and in #27511, and I understand the concerns around compatibility.
The filtered view (based on `IsTerrible()`) may be misleading for users who assume those are the only addresses known — unless they're explicitly inspecting address table quality.
That said, happy to follow the current direction — just sharing those thoughts for context.

<details>
<summary>Comparing this latest output against my <a href="https://github.com/bitcoin/bitcoin/pull/26988#pullrequestreview-1679047963">previous testing</a> I noticed that the <code>all_networks</code> element has been removed (?).</summary>

```
./build_26988/bin/bitcoin-cli -signet -datadir=/tmp/btc -addrinfo
{
  "all addresses known (used for selecting peers)": {
    "ipv4": 1368,
    "ipv6": 893,
    "onion": 0,
    "i2p": 0,
    "cjdns": 0,
    "total": 2261
  },
  "addresses known after quality/recency filtering (for original -addrinfo compatibility)": {
    "ipv4": 1352,
    "ipv6": 883,
    "onion": 0,
    "i2p": 0,
    "cjdns": 0,
    "total": 2235
  }
}
```

</details>

jonatack commented on doc/release-notes-26988.md:
```diff
@@ -0,0 +1,6 @@
+Tools and Utilities
+--------
+
+- CLI -addrinfo previously (v22 - v29) returned addresses known to the node after filtering for quality and recency.
```
```suggestion
- CLI -addrinfo previously (v22 - v30) returned addresses known to the node after filtering for quality and recency.
```

jonatack reviewed (commented):
Approach ACK 016ab85a13408ad980c3dbce4e041e14a4fcf3b8, thanks for updating, tested with a server on master, not yet tested with a server running pre-v26 (before getaddrmaninfo)

```
$ ./build/bin/bitcoin-cli -addrinfo
{
  "all addresses known (used for selecting peers)": {
    "ipv4": 48133,
    "ipv6": 8017,
    "onion": 17257,
    "i2p": 3757,
    "cjdns": 11,
    "total": 77175
  },
  "addresses known after quality/recency filtering (for original -addrinfo compatibility)": {
    "ipv4": 45921,
    "ipv6": 7487,
    "onion": 16259,
    "i2p": 3035,
    "cjdns": 7,
    "total": 72709
  }
}
```

vasild commented on src/bitcoin-cli.cpp:
```diff
@@ -274,19 +274,49 @@ struct BaseRequestHandler {
+                addresses.pushKV(network_types[i], addr_count);
+            }
+            uint64_t total = addrman_counts[network_types.size() - 1]["total"].getInt<int>();
+            addresses.pushKV("total", total);
```
This assumes that the `all_networks` key is the last key in the JSON returned by `getaddrmaninfo` but JSON objects don't have order (unlike arrays). So it is possible that the order changes any time (in theory) and then this code will break. Better write it in such a way that it works regardless of the order in which keys are returned.

Maybe something like:
```
iterate all from i = 0 to i < network_types.size()
    if key is all_networks
        this is the total
    else
        this is one of the networks

vasild reviewed (commented):
Reviewed as of 016ab85a13

stratospher commented:
thank you for the reviews! I've pushed an [update](https://github.com/bitcoin/bitcoin/compare/016ab85a13408ad980c3dbce4e041e14a4fcf3b8..858f49bcb3590b211f014e8d290008bb13f5b17f) changing the long key names.
also retained the original key name in -addrinfo's result (`addresses_known`) in case some other scripts/tools rely on it. 
```
$ build/bin/bitcoin-cli -addrinfo
{
  "addresses_known": {
    "ipv4": 52037,
    "ipv6": 8855,
    "onion": 3465,
    "i2p": 0,
    "cjdns": 0,
    "total": 64357
  },
  "addresses_filtered": {
    "ipv4": 41811,
    "ipv6": 5808,
    "onion": 2465,
    "i2p": 0,
    "cjdns": 0,
    "total": 50084
  }
}
```

regarding https://github.com/bitcoin/bitcoin/pull/26988#issuecomment-3477855579:
> 25.x, 26.x and 27.x are all end-of-life; so that seems out of scope for needing to test / accomodate code-wise?

Good point. Thinking of changing it in the next push.

According to bitnodes.io - among bitcoin core nodes, 95% are already on v26+ and therefore support getaddrmaninfo (introduced 2 years ago).

Maybe the concern regarding older node <-> newer CLI compatibility isn't there anymore considering how much the network/software has changed! (cc @jonatack, curious to hear your thoughts and anyone else's of course)
- 95% of bitcoin core nodes are on v26+ and have access to getaddrmaninfo (introduced 2 years ago)
- the nodes we're concerned about breaking user space (v22 - v26 nodes) with have reached EOL

fanquake commented:
> CLI calls can be made to a node running a different, older version. If a researcher, data gatherer, developer, or node operator is calling this on a pre-getaddrmaninfo node, 

Could you elaborate on why these people, particularly the developers, or node operators, are running EOL versions of Core (with known CVEs); and why they can't they upgrade their nodes, but at the same time, they are downloading and using the latest version of bitcoin-cli?

> There doesn't seem to be any reason not to do the same here, rather than choosing to break it.

The reason to not do it, is that it then implies we are still supporting versions of our software, which we have marked as EOL.

mzumsande reviewed (commented):
> CLI calls can be made to a node running a different, older version. If a researcher, data gatherer, developer, or node operator is calling this on a pre-getaddrmaninfo node, then the call should still work. As commented above, colleagues here requested that -netinfo continue to work in such cases, and we patched it. There doesn't seem to be any reason not to do the same here, rather than choosing to break it.

I don't agree with that. I don't see `bitcoin-cli` and `bitcoind` as completely separate programs that should be supported in arbitrary combinations for eternity.

If that was the case, it would have been a bad mistake to ever have added non-trivial functionality like `-addrinfo` or `-netinfo` to `bitcoin-cli` instead of `bitcoind` in the first place because that would have created an infinite obligation to maintain compatibility of current `bitcoind` versions with ancient versions of `bitcoin-cli` long out of support.

While it makes some limited sense to try to keep up compatibility over version ranges supported at the same time (although even that is debatable in my opinion), I don't think that it is reasonable to make any efforts supporting compatibility beyond this.

vasild reviewed (approved):
ACK 5b05a9959f1633bfee78d9edb180c672b0640ab5

pablomartin4btc reviewed (commented):
tACK 5b05a9959f1633bfee78d9edb180c672b0640ab5

with `bitcoind` < v26:
```
/build/bin/bitcoin-cli -signet -datadir=/tmp/btc -addrinfo
error: -addrinfo requires bitcoind v26.0 or later which supports getaddrmaninfo RPC. Please upgrade your node or use bitcoin-cli from the same version.

```

vasild reviewed (approved):
ACK 325f98bca07c3a0c48a46dc18f67376d44c5341d

pablomartin4btc reviewed (commented):
ACK 325f98bca07c3a0c48a46dc18f67376d44c5341d

_(Left a small [suggestion](https://github.com/bitcoin/bitcoin/pull/26988/files#r2615258470) since the older node may be operated by someone other than the CLI user, and upgrading it might not be possible or appropriate when this change is the only reason. I think this would also align with @jonatack’s earlier [note](https://github.com/bitcoin/bitcoin/pull/26988#issuecomment-3559469161).)_

pablomartin4btc reviewed (approved):
re-ACK b3046cca7182f3399a221757318d24e203092301

vasild reviewed (approved):
ACK b3046cca7182f3399a221757318d24e203092301

sr-gi reviewed (commented):
tACK b3046cca7182f3399a221757318d24e203092301

I tested this on top of b3046cca7182f3399a221757318d24e203092301 and against a 25.2 and 26.0 instances. I can confirm the exception is thrown for version < v26.
